irohad/ametsuchi/impl/postgres_ordering_service_persistent_state.cpp

```diff
@@ -0,0 +1,132 @@
+          std::make_unique<pqxx::lazyconnection>(postgres_options);
+      try {
+        postgres_connection->activate();
+      } catch (const pqxx::broken_connection &e) {
+        return expected::makeError(
+            (boost::format("Connection to PostgreSQL broken: {}") % e.what())
+                .str());
+      }
+      log_->info("connection to PostgreSQL completed");
+
+      // create transaction
+      auto postgres_transaction = std::make_unique<pqxx::nontransaction>(
+          *postgres_connection, "Storage");
+      log_->info("transaction to PostgreSQL initialized");
+
+      expected::Result<std::shared_ptr<PostgresOrderingServicePersistentState>,
+                       std::string>
+          storage;
+      storage = expected::makeValue(
+          std::shared_ptr<PostgresOrderingServicePersistentState>(
+              new PostgresOrderingServicePersistentState(
+                  std::move(postgres_connection),
+                  std::move(postgres_transaction))));
+      return storage;
+    }
+
+    PostgresOrderingServicePersistentState::
+        PostgresOrderingServicePersistentState(
+            std::unique_ptr<pqxx::lazyconnection> postgres_connection,
+            std::unique_ptr<pqxx::nontransaction> postgres_transaction)
+        : postgres_connection_(std::move(postgres_connection)),
+          postgres_transaction_(std::move(postgres_transaction)),
+          log_(logger::log("PostgresOrderingServicePersistentState")),
+          execute_{ametsuchi::makeExecuteResult(*postgres_transaction_)} {}
+
+    void PostgresOrderingServicePersistentState::initStorage() {
+      log_->info("Init storage");
+      postgres_transaction_->exec(
+          "CREATE TABLE IF NOT EXISTS ordering_service_state (\n"
+          "    proposal_height bigserial\n"
+          ");"
+          "INSERT INTO ordering_service_state\n"
+          "VALUES (2); -- expected height (1 is genesis)");
+    }
+
+    void PostgresOrderingServicePersistentState::dropStorgage() {
+      log_->info("Drop storage");
+      postgres_transaction_->exec(
+          "DROP TABLE IF EXISTS ordering_service_state;");
+    }
+
+    bool PostgresOrderingServicePersistentState::saveProposalHeight(
+        size_t height) {
+      log_->info("Save proposal_height in ordering_service_state "
+                 + std::to_string(height));
+      return execute_(
+                 "DELETE FROM ordering_service_state;\n"
+                 "INSERT INTO ordering_service_state "
+                 "VALUES ("
+                 + postgres_transaction_->quote(height) + ");")
+          .match([](expected::Value<pqxx::result> v) -> bool { return true; },
+                 [&](expected::Error<std::string> e) -> bool {
+                   log_->error(e.error);
+                   return false;
+                 });
+    }
+
+    boost::optional<size_t>
+    PostgresOrderingServicePersistentState::loadProposalHeight() const {
+      boost::optional<size_t> height;
```

Comment:
Is boost::variant conducts an initialization of objects? If no, need to set value explicitly.

Reply:
boost::optional creates uninitialized value that will be initialized further or not